age-plugin.md

```diff
@@ -0,0 +1,541 @@
+{%hackmd theme-dark %}
+
+# Plugin system for age
+
+## Introduction
+
+The age file encryption format follows the "one well-oiled joint" design
+philosophy. The mechanism for extensibility (within a particular format version)
+is the recipient stanzas within the age header: file keys can be wrapped in any
+number of ways, and age clients are required to ignore stanzas that they do not
+understand. The core APIs that exercise this mechanism are:
+
+- A recipient that wraps a file key and returns a stanza.
+- An identity that unwraps a stanza and returns a file key.
+
+Custom age clients can bundle support for the exact recipient or identity types
+required for a particular environment. However, a general plugin system will
+enable an ecosystem of common third-party recipient types to grow.
+
+The plugin system specified in this document provides a mechanism for exposing
+the core APIs across process boundaries. It has two main components:
+
+- A map from recipients and identities to plugin binaries.
+- State machines for wrapping and unwrapping file keys.
+
+With this composable design, developers can implement a recipient or identity
+that they would use directly with an age library, and then also deploy it as a
+plugin binary.
+
+## Mapping recipients and identities to plugin binaries
+
+age plugins are identified by an arbitrary case-insensitive string `NAME`. This
+string is used in three places:
+
+- Plugin-compatible recipients are encoded using Bech32 with the HRP `age1name`
+  (lowercase).
+- Plugin-compatible identities are encoded using Bech32 with the HRP
+  `AGE-PLUGIN-NAME-` (uppercase).
+- Plugin binaries (to be started by age clients) are named `age-plugin-name`.
```

Comment:
Custom stanzas are not required to use the same `name` as the plugin, so that consideration is separate. For example, my `age-plugin-yubikey` plugin defined the stanza `piv-p256` (named without FQDN because it predates the age spec recommendation).

As for having both 1 and 2, that's a usability decision: hyphenated binary names are easier to read, and having `plugin` in the name helps with domain separation in the binary name space. But they aren't selectable with a double-click, which is something we wanted for recipient strings.